File: argon-backend/js-node/src/main/scala/dev/argon/backend/sjs/BackendFactory.scala
package dev.argon.backend.sjs

import dev.argon.backend.sjs.metadata.BackendMetadata
import nobleidl.sjs.core.ErrorChecker

import scala.scalajs.js

trait SimpleBackendFactory extends js.Object {
  def create[E, A](errorChecker: ErrorChecker[E], hostOperations: HostOperations[E], f: js.Function1[Backend[E, ?], A]): A
}

trait BackendFactory extends SimpleBackendFactory {
  def metadata: BackendMetadata
  def close(): js.Promise[Unit]
}

File: backend/backends/java/api/src/main/java/dev/argon/backend/api/BackendFactory.java
package dev.argon.backend.api;

import dev.argon.nobleidl.runtime.ErrorType;

public interface BackendFactory {
	<TE, EE extends Throwable> Backend<TE, ?> create(ErrorType<TE, EE> errorType, HostOperations<TE> hostOperations) throws InterruptedException;
}
